src/features/cart/ShoppingCartTest.ts

```diff
@@ -0,0 +1,185 @@
+            }
+            else{
+                console.log("This Code allready Exits in the Cart",isApplied)
+                return
+            }
+        }
+        else{
+            const product=cart.items[cart.items.findIndex((item)=>item.id==checkCode.itemId)]
+            //All ready Use or not
+            const isApplid=product.discount.includes(checkCode.code)
+            if(!isApplid){
+                product.discount.push(checkCode.code)
+                console.log("Descount Code is item",product)
+
+            }else{
+                console.log("All Ready Use this code before")
+                return
+            }
+
+        }
+
+    }
+    else{
+        console.log("There Is No Code Here ",discountCode)
+    }
+
+}
+//Get Single Card Price 
+export const singlePrice=(item:Item):number=>{
+    const codeApplied = item.discount.length > 0;
+  let totalPrice = 0;
+  if (codeApplied) {
+    // get code and then its value from discountCodes
+    const codeValue = discountCodes.filter((code) =>
+      item.discount.includes(code.code)
+    );
+    console.log(codeValue);
+    const totalDiscount = codeValue.reduce(
+      (total, item) => total + item.value,
+      0
+    );
+    console.log(totalDiscount);
+    totalPrice =(item.price * item.quantity) - ((item.price * item.quantity * totalDiscount) / 100);
+  } else {
+    totalPrice = item.quantity * item.price;
+  }
+  return totalPrice;
+}
+export const grandTotal=():number=>{
+  // check if any item has discounted price 
+  const haveDiscount = cart.items.filter(item => item.discount.length > 0);
+  let totalDiscountedItemPrice : number = 0;
+  if(haveDiscount){
+    const discountedItemPrice : number[] = [];
+    // get discounted price of these items 
+    haveDiscount.forEach(item => {
+     const price =  singlePrice(item);
+     discountedItemPrice.push(price)
+    })
+    totalDiscountedItemPrice = discountedItemPrice.reduce((total, item) => total + item, 0)
+  }
+  const codeApplied = cart.discount.length > 0;
+  let totalPrice = 0;
+  if (codeApplied) {
+    const codeValue = discountCodes.filter((code) =>
+      cart.discount.includes(code.code)
+    );
+    const totalDiscount = codeValue.reduce(
+      (total, item) => total + item.value,
+      0
+    );
+    const cartPrice = cart.items.reduce((total, item) => {
+      if(item.discount.length === 0){
+        return total + item.quantity * item.price
+      } else {
+        return total
+      }
+    }, 0);
+    const finalPrice = cartPrice + totalDiscountedItemPrice;
+
+    totalPrice = finalPrice - (finalPrice * totalDiscount / 100);
+  } else {
+    const cartPrice = cart.items.reduce((total, item) => {
+      if(item.discount.length === 0){
+        return total + item.quantity * item.price
+      } else {
+        return total
+      }
+    }, 0);
+    totalPrice = cartPrice + totalDiscountedItemPrice;
+    
+  }
+  return totalPrice;
+}
+// show all cart items
```

Comment:
It would be great if you could add comments to these functions using proper JSDoc syntax.